### PR TITLE
RTCRtpSender and MediaStream support

### PR DESCRIPTION
**Pull request by markandrus**
_Not ready to merge._

See checklist in #396.

One thing to figure out how to do: raising `onaddtrack` and `onremovetrack` on the MediaStream. So far, objects that raise events extend from EventLoop; but these need to be stopped via Stop. There's no clear event that should trigger Stop, at least for both the local and remote MediaStreams. For remote MediaStreams this could coincide with closing the RTCPeerConnection; but I'm not sure about local MediaStreams. Perhaps MediaStream's private constructor could accept a boolean indicating whether it is local or not, and then we let local MediaStreams stop themselves immediately? That could solve the problem, although it feels a little weird. I think the alternative would require RTCPeerConnection driving events on remote MediaStreams it "owns" itself. The annoying thing, though, is that the observer for the MediaStream would get registered on the RTCPeerConnection rather than the MediaStream, and then the RTCPeerConnection would have to reach into the MediaStream to raise the events. Perhaps the former is cleaner then.

### CODE REVIEW

File: src/converters.h [hunk 1]
```diff
@@ -63,8 +63,8 @@ struct Converter<T, T> {
 template <typename S, typename L, typename R>
 struct Converter<S, Either<L, R>> {
   static Validation<Either<L, R>> Convert(const S s) {
-    return From<L>(s).Map(&Either<L, R>::Left)
-        | (From<R>(s).Map(&Either<L, R>::Right));
+    return Converter<S, L>::Convert(s).Map(&Either<L, R>::Left)
+        | (Converter<S, R>::Convert(s).Map(&Either<L, R>::Right));
```
Comment (markandrus) [inline src/converters.h]: I hit some issues with ambiguous template specialization. Might be able to revert to the nicer-to-read version.

File: src/converters/arguments.h [hunk 2]
```diff
@@ -23,10 +23,25 @@
 
 namespace node_webrtc {
 
+struct Arguments {
+ public:
+  Nan::NAN_METHOD_ARGS_TYPE info;
+  explicit Arguments(Nan::NAN_METHOD_ARGS_TYPE info): info(info) {}
+};
```
Comment (markandrus) [inline src/converters/arguments.h]: I hit some issues with ambiguous template specialization. Might be able to revert this change.

File: src/converters/webrtc.h [hunk 3]
```diff
@@ -620,6 +623,48 @@ struct Converter<webrtc::MediaStreamTrackInterface::TrackState, std::string> {
...
+
+template <>
+struct Converter<v8::Local<v8::Value>, node_webrtc::MediaStreamTrack*> {
+  static Validation<node_webrtc::MediaStreamTrack*> Convert(v8::Local<v8::Value> value) {
+    // TODO(mroberts): This is not safe.
+    return value->IsObject() && !value->IsArray()
+        ? Validation<node_webrtc::MediaStreamTrack*>::Valid(Nan::ObjectWrap::Unwrap<node_webrtc::MediaStreamTrack>(value->ToObject()))
+        : Validation<node_webrtc::MediaStreamTrack*>::Invalid("IDK");
+  }
+};
```
Comment (markandrus) [inline src/converters/webrtc.h]: The safe way to do this, and the conversion above, would be something like this: https://stackoverflow.com/questions/8994196/how-to-check-for-correct-type-when-calling-objectwrapunwrap-in-a-nodejs-add-on

File: src/mediastream.cc [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+  auto self = Nan::ObjectWrap::Unwrap<MediaStream>(info.Holder());
+  auto tracks = std::vector<MediaStreamTrack*>();
+  for (auto const& track : self->_stream->GetAudioTracks()) {
+    auto mediaStreamTrack = MediaStreamTrack::GetOrCreate(self->_factory, track);
+    tracks.push_back(mediaStreamTrack);
+  }
+  for (auto const& track : self->_stream->GetVideoTracks()) {
+    auto mediaStreamTrack = MediaStreamTrack::GetOrCreate(self->_factory, track);
+    tracks.push_back(mediaStreamTrack);
+  }
```
Comment (markandrus) [inline src/mediastream.cc]: This is used in a couple places. Worth factoring out into a `tracks` method. Then we can reuse in `GetTracks`, `GetActive`, `GetTrackById`, etc.

File: src/mediastream.cc [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+  if (track->kind() == track->kAudioKind) {
+    stream->RemoveTrack(static_cast<webrtc::AudioTrackInterface*>(track.get()));
+  } else {
+    stream->RemoveTrack(static_cast<webrtc::VideoTrackInterface*>(track.get()));
+  }
+}
+
+NAN_METHOD(MediaStream::Clone) {
+  Nan::ThrowError("Not yet implemented; file a feature request against node-webrtc");
+}
```
Comment (markandrus) [inline src/mediastream.cc]: Implement this by returning the C++ equivalent of `new MediaStream(this)`.

File: src/mediastream.h [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+ public:
+  MediaStream(
+      std::shared_ptr<node_webrtc::PeerConnectionFactory>&& factory,
+      rtc::scoped_refptr<webrtc::MediaStreamInterface>&& stream):
+    MediaStream(std::move(factory), std::move(stream), !_shouldReleaseFactory) {}
+
+  MediaStream(
+      std::shared_ptr<node_webrtc::PeerConnectionFactory>&& factory,
+      rtc::scoped_refptr<webrtc::MediaStreamInterface>&& stream,
+      bool shouldReleaseFactory);
```
Comment (markandrus) [inline src/mediastream.h]: Would be good to comment on these. The reason you might pass a non-null `factory` _and_ set `shouldReleaseFactory` to `true` are a result of the `new MediaStream(tracks)` constructor and needing to use the default factory. Actually, maybe we could still encapsulate this logic inside the "smaller" constructor when `factory` and `stream` are `nullptr`.

**Comment by aldahick** [comment]
Forgive me if this is obvious, but I'm not sure what the progress on this implementation is. Is it still being blocked by `onaddtrack` and `onremovetrack`? As many other people have mentioned in other places, this will be great once it's published!

**Comment by markandrus** [comment]
Hey @aldahick,

I saw a crash for one of the Appveyor builds. I want to try to replicate in a Windows VM and fix it before merging. I’ve just been busy recently! I think I’ll have some time this week/weekend.

**Comment by markandrus** [comment]
I looked into the bug. It's reproducible on Linux on macOS, too, in Node 9.8.0, AFAICT. I'm not sure if it's a Nan bug or something wrong with the way I'm using Nan/v8 and libuv APIs. I created a minimal repro example [here](https://github.com/markandrus/nan-asyncresource-bug) and opened up an issue on nodejs/nan.

**Comment by markandrus** [comment]
For reference, the bug I filed is here: https://github.com/nodejs/nan/issues/772 **EDIT**: I merged #407 which gives a new pattern for handling AsyncResource + ObjectWrap and rebased this branch.

**Comment by markandrus** [comment]
Answering my own questions:

> One thing to figure out how to do: raising `onaddtrack` and `onremovetrack` on the MediaStream.

`onaddtrack` and `onremovetrack` will only be raised on remote MediaStreams. the PeerConnection class, when handling the `ontrack` event, can raise the event on MediaStream.

However, I would like to punt on `onaddtrack` and `onremovetrack` for now so that I can merge this PR sooner than later. In my testing, things seem pretty good. I don't believe there are any memory leaks with the new objects. We also have destructor tests now to ensure there are no leaks.
